src/util/cow.zig

```diff
@@ -0,0 +1,169 @@
+const std = @import("std");
+const Allocator = std.mem.Allocator;
+
+const assert = @import("../util.zig").assert;
+const debugAssert = @import("../util.zig").debugAssert;
+const IS_DEBUG = @import("../util.zig").IS_DEBUG;
+
+/// A clone-on-write type monomorphized for strings.
+///
+/// I tried adding this to smart-pointers but it turned out to be a pain to
+/// generalize.  I need this structure _right now_; we should go back and
+/// move+generalize this code later.
+pub fn Cow(comptime sentinel: bool) type {
+    const MutSlice = if (sentinel) [:0]u8 else []u8;
+    const Slice = if (sentinel) [:0]const u8 else []const u8;
+
+    return struct {
+        /// Does this `Cow` own its data, or is it borrowing it from someone
+        /// else? `true` if borrowed.
+        ///
+        /// **HEY YOU**. Yeah, you. Are you using this struct externally? If so,
+        /// _DON'T MUTATE THIS FIELD_!
+        borrowed: bool,
```

Comment:
if that's true, why not prefix it with `_`?

Reply:
It can be read by external consumers, just not modified